ansibullbot/triagers/plugins/docs_info.py

```diff
@@ -0,0 +1,306 @@
+
+    if source is not None:
+        for line in diff:
+            # Check if change applies to module DOCUMENTATION (if exists)
+            if source.doc_string:
+                if source.ds_line_start <= line["lineno"] <= source.ds_line_end:
+                    continue
+
+            # Check if change applies to module EXAMPLES (if exists)
+            if source.example_string:
+                if source.ex_line_start <= line["lineno"] <= source.ex_line_end:
+                    continue
+
+            # Find appropriate class if it exists
+            source_class = source.find_class(line["lineno"])
+            if source_class is not None:
+                if source_class.doc_string:
+                    if source_class.ds_line_start <= line["lineno"] <= source_class.ds_line_end:
+                        continue
+
+                source_func = source_class.find_function(line["lineno"])
+                if source_func is not None and source_func.doc_string:
+                    if source_func.ds_line_start <= line["lineno"] <= source_func.ds_line_end:
+                        continue
+
+            # If we made it this far, this change is outside docs/examples
+            return False
+
+        return True
+
+def _is_docs_only(changed_file):
+    """ Check if the changes made to ``changed_file`` affect only documentation. """
+    docs_only = False
+
+    if isinstance(changed_file, dict):
+        changed_file = CommitFile(changed_file)
+
+    if _is_docs_path(changed_file.filename):
+        docs_only = True
+    else:
+        # Additions or deletions of complete files outside of
+        # docsite/examples will likely be more than documentation
+        # changes
+        if changed_file.status != "modified":
+            docs_only = False
+        else:
+            # Non-Python files that are outside of the docsite
+            # folder will not be documentation related
+            if not changed_file.filename.endswith(".py"):
+                docs_only = False
+            else:
+                # If a python file, check if the changes are only to
+                # docstrings
+                if _check_py_changes(changed_file.file_content, changed_file.patch):
+                    docs_only = True
+
+    return docs_only
+
+def get_docs_facts(iw):
+    """ Cycle through the files and gather facts about documentation changes. """
+    dfacts = {
+        "is_docs_only": False
+    }
+
+    if not iw.is_pullrequest():
+        return dfacts
+
+    docs_only = False
+
+    for commit in iw.commits:
+        commit_files = iw.get_commit_files(commit)
+        if commit_files is None:
+            # "Sorry, this diff is temporarily unavailable due to heavy server load."
+            # Preserve docsite label to prevent potential waffling
+            dfacts["is_docs_only"] = "docs_only" in iw.labels
+            return dfacts
+
+        has_non_docs_changes = any(filterfalse(_is_docs_only, commit_files))
```

Comment:
I think this is too clever and the readability would be better with classic `for` loop here - the combination of `has_non_docs_changes`, `any`, `filterfalse` and `_is_docs_only` confused me at first. Just a note, I think I am fine with staying this as is though.

Reply:
I have had the same thoughts, while writing it and have been stewing on it. I'll work some more iterations to improve readability.

Reply:
So two things on this:

1. I realize now that I had tunnel vision on finding falsey cases, which lead to the cleverness.
2. I re-visited `issuewrapper.get_files()` from https://github.com/ansible/ansibullbot/pull/1505#discussion_r574552256. When I looked at it before, I made my conclusion based only on the repr. It actually has everything we need, and would allow to only focus on the changed content and not each commit. I'm still not sure if the botmeta desires the by-commit-history, though. With `get_files()` and can actually get this down to one or two lines:
   ```python
    pr_files = iw.get_files()
    docs_only = all([_is_docs_only(f.raw_data) for f in pr_files])

    # OR (without all(); explicitly looking for False)
    pr_files = iw.get_files()
    docs_only = False not in [_is_docs_only(f.raw_data) for f in pr_files]
   ```
   This would drive a fix to `get_files()`, as it currently tries to set a non-existent attribute (`.files`). Not sure if it should work _with_ the `files` property, or separately. Would also drive more changes to `IssueMock` for testing, but that is trivial.

Reply:
> I'm still not sure if the botmeta desires the by-commit-history, though.

Can you elaborate on the above? What do you mean by `botmeta`?

Reply:
> Can you elaborate on the above? What do you mean by botmeta?

This derived from my uncertainty in how the meta/history worked with cached info. I was afraid that it _needed_ the history to be ordered by commits.

I took some more time to study the botmeta & history caching. I understand now that it is all cached by the issue/PR, so using the commits isn't necessarily required.